### ISSUE TITLE
Enhance Animations within Stories Editor to Include Additional Supported Animation Properties

### ISSUE THREAD

**Issue by jauyong** [ampproject/amp-wp#3318]
## Feature description

As a user I want to be able to have more animation options than currently exist. There are supported animation features that are not yet implemented that would enhance the animation experience.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* AC1: Implement the, already supported, ability to specify timing function within an animation.
* AC2: Implement, the already supported, scaling animation enhancement 
* AC3: Implement, the already supported, translating animation enhancement.

Related to #3421 and #2533 
 
### Timing function

This adds support for setting [`animate-in-timing-function` as supported by AMP](https://amp.dev/documentation/components/amp-story/#animate-in-timing-function-[optional]).

Add a new field to every element, that can be animated, called `timingValues`. This will either be `null` or an array of exactly 4 numbers each between 0 and 1.

Add a new control below the current delay control called "Timing function". It will be a drop down, with _"Linear"_ selected as default. The other options are _"Ease in", _"Ease out"_ and _"Ease in-out"_. For the non-linear functions, the following lists of timing values will be used:

* *"Ease in"*: `0.42, 0, 1, 1`.
* *"Ease out"*: `0, 0, 0.58, 1`.
* *"Ease in-out"*: `0.42, 0, 0.58, 1`.

This will be stored as `timingValues` on the element as the 4 value list or `null`. For _"Linear"_ and as default, `null` will be stored in `timingValues`.

When rendered, `null` should be ignored and any valid list of 4 numbers between 0 and 1 should be rendered as `animate-in-timing-function="cubic-bezier(timingValues[0], timingValues[1], timingValues[2], timingValues[3])"`.

### Optional properties.

Add a new field to every element, that can be animated, called `optionalValues`. This will be an object with keys relevant only for specific animation types. If the object has values unused by the current animation type, they will be ignored.

#### Zoom

This adds support for setting [`scale-start` and `scale-end` as supported by AMP](https://amp.dev/documentation/components/amp-story/#scale-start,-scale-end-[optional,-only-works-with-zoom-in-&-zoom-out-animations]).

If the user selects either the animation type _"Zoom in"_ or _"Zoom out"_, add two range controls (with input fields next to them) just above the _"Play animation"_ button for "Initial zoom" and "Final zoom" values.

The ranges should go from 0.1 to 10 with 0.1 increments.

The default values are as follows:

* **Zoom in**: _Initial zoom_: 1, _Final zoom_: 3.
* **Zoom out**: _Initial zoom_: 3, _Final zoom_: 1.

There should be a "reset to default" button below both ranges, that will reset to the above defaults.

Store these if set to non-default values on the `optionalValues` object as `zoomStart` and `zoomEnd` respectively. If either value is set to default, make sure no corresponding property is set on the `optionalValues` object.

Render these if set for the right type of animation as `scale-start="optionalValues.zoomStart" scale-end="optionalValues.zoomEnd"`.

#### Horizontal pan

This adds support for setting [`translate-x` as supported by AMP](https://amp.dev/documentation/components/amp-story/#translate-x-[optional,-only-works-with-pan-left-&-pan-right-animations]).

If the user selects either the animation type _"Pan left"_ or _"Pan right"_, add a range control (with input field next to it) just above the _"Play animation"_ button for a "Horizontal distance" value in percent of the element width.

The range goes from 0 to 100% and the default value is 100%. There's a button "reset to default", that will reset the range to 100%.

This value is displayed to editors as 0-100%, but actually stored as a scaled value the block animation meta data. The scaled value is different for _"Pan Left"_ and _"Pan Right"_ as defined below:

```js
const dimensions = { 
	pageWidth, // a constant
	pageHeight, // a constant
	targetWidth, // image original width, not actual width
	targetHeight, // image original height, not actual height
};
const scalingFactor = calculateTargetScalingFactor( dimensions );
const offsetX = dimensions.pageWidth - ( dimensions.targetWidth * scalingFactor );

/**
 * @param {Number} percentage A value between 0 and 100 with 0 corresponding to no pan and 100 to maximum pan.
 * @return {Number} The desired pan in pixels.
 */
const getPanLeftTranslateX = ( percentage ) => {
	return ( percentage - 100 ) * offsetX;
};

/**
 * @param {Number} percentage A value between 0 and 100 with 0 corresponding to no pan and 100 to maximum pan.
 * @return {Number} The desired pan in pixels.
 */
const getPanRightTranslateX = ( percentage ) => {
	return percentage * offsetX;
};
```

With [`calculateTargetScalingFactor` already in `helpers`](https://github.com/ampproject/amp-wp/blob/develop/assets/src/stories-editor/helpers/setAnimationTransformProperties.js#L31-L44).

`scalingFactor` and thus `offsetX` is static based on the original image and doesn't change as the image is manipulated on page.

Store this if set as `optionalValues.translateX = getPanLeftTranslateX( value )` or `optionalValues.translateX = getPanRightTranslateX( value )` respectively. If slider is set to 100 (default), there should be no `translateX` on the `optionalValues` object.

Render this if set for the right type of animation as `translate-x="optionalValues.translateX"`.

#### Vertical pan

This adds support for setting [`translate-y` as supported by AMP](https://amp.dev/documentation/components/amp-story/#translate-y-[optional,-only-works-with-pan-up-&-pan-down-animations]).

If the user selects either the animation type _"Pan up"_ or _"Pan down"_, add a range control (with input field next to it) just above the _"Play animation"_ button for a "Vertical distance" value in percent of the element height.

The range goes from 0 to 100% and the default value is 100%. There's a button "reset to default", that will reset the range to 100%

This value is displayed to editors as 0-100%, but actually stored as a scaled value the block animation meta data. The scaled value is different for _"Pan Up"_ and _"Pan Down"_ as defined below:

```js
const dimensions = { 
	pageWidth, // a constant
	pageHeight, // a constant
	targetWidth, // image original width, not actual width
	targetHeight, // image original height, not actual height
};
const scalingFactor = calculateTargetScalingFactor( dimensions );
const offsetY = ( dimensions.pageHeight - ( dimensions.targetHeight  * scalingFactor ) ) / 2;

/**
 * @param {Number} percentage A value between 0 and 100 with 0 corresponding to no pan and 100 to maximum pan.
 * @return {Number} The desired pan in pixels.
 */
const getPanUpTranslateY = ( percentage ) => {
	return ( percentage - 100 ) * offsetY;
};

/**
 * @param {Number} percentage A value between 0 and 100 with 0 corresponding to no pan and 100 to maximum pan.
 * @return {Number} The desired pan in pixels.
 */
const getPanDownTranslateY = ( percentage ) => {
	return percentage * offsetY;
};
```

With [`calculateTargetScalingFactor` already in `helpers`](https://github.com/ampproject/amp-wp/blob/develop/assets/src/stories-editor/helpers/setAnimationTransformProperties.js#L31-L44).

`scalingFactor` and thus `offsetY` is static based on the original image and doesn't change as the image is manipulated on page.

Store this if set as `optionalValues.translateY = getPanUpTranslateY( value )` or `optionalValues.translateY = getPanDownTranslateY( value )` respectively. If slider is set to 100 (default), there should be no `translateY` on the `optionalValues` object.

Render this if set for the right type of animation as `translate-y="optionalValues.translateY"`.

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by miina** [comment]
I was looking through what the allowed animation options are and looks like there are some that we haven't implemented yet:
- `animate-in-timing-function`
- `scale-start, scale-end` [only works with zoom-in & zoom-out animations]
- `translate-x` / `translate-y` [optional, only works with pan-left & pan-right / pan-up & pan-down animations]
- Combining multiple animations

Combining multiple animations could be interesting, more than one animation is not allowed per element, however, the recommended way seems to be wrapping the elements, e.g.
```
<div animate-in="fly-in-left">
   <div animate-in="fade-in">
     I will fly-in and fade-in!
   </div>
</div>
```

One option could be starting with those enhancements, for example we could add support for combining multiple animations where additionally there could be a choice of if to play the added animation at the same time or after the previous. Needs testing if that's possible, no reason so far why it shouldn't be.

For starters, just one additional combined animation could be an option.

Thoughts?

cc @swissspidy

**Comment by swissspidy** [comment]
Are these attributes all supported in AMP Stories too? I thought AMP Stories is a bit more strict there.

Nesting animations is doable, but also adds quite a bit of complexity to the whole thing. Needs some exploring and comparison with other editors I think to see how it can be done and whether it is deemed useful.

For things like #2533 it would probably easier to just use custom CSS animations instead of nesting.

**Comment by miina** [comment]
> Are these attributes all supported in AMP Stories too? I thought AMP Stories is a bit more strict there.

Yes, this is taken from the AMP Stories docs: https://amp.dev/documentation/components/amp-story/?format=websites#combining-multiple-animations

Perhaps nesting animations could make sense dynamically (PHP).

> Needs some exploring and comparison with other editors I think to see how it can be done and whether it is deemed useful.

👍

**Comment by swissspidy** [comment]
Ha, looks like I even created an issue months ago for these attributes: #2447. Will close it as a duplicate.

Also interesting: #2529 for animation playback UI (timeline)

**Comment by miina** [comment]
> Could we perhaps have only 1 animations block attribute which contains an array of animations? We'd then handle those in PHP, with the outermost element having the id that can be referenced by other blocks with animate-in-after. In the UI, animations would be a "repeatable" field.

Yep, I was also thinking of just one field with repeatable UI select. Changing the current field will probably involve migration, too. Since currently it's in JS. Will need to investigate a bit to see how useful this could be.

**Comment by miina** [comment]
Note: It was confirmed that or now we can continue with the already supported animation enhancements.

Will create an issue for that with details shortly. We might want to have a separate issue for combining animations not to bloat the PR.

**Comment by spacedmonkey** [comment]
I find implementation brief extremely hard to follow without driving extremely deepingly on this subject area. I have very simple questions that are not answered anywhere?

- What are the new animations that going to be added to the existing list of supported? Please provide a list. 
- What new requirements do these new animations require, such as new input fields and stored data?
- Will existing animations be effected? Will a migration be required?
- What is required to implement these animations in the editor?
- How will data be stored? At block level, post meta etc. 
- What requires are they for accessibility? 
    - Has they been a accessibility review on `BezierEditor` component?
    - Could existing gutenberg components be used?
- Default values are unknowns at the moment, how can we make them unknown?

I think that @miina might be a better person to review this, as she has a deeper understanding of the subject matter.

**Comment by miina** [comment]
Thanks for the note, @spacedmonkey, will review this shortly as well.

**Comment by barklund** [comment]
@spacedmonkey, as far as I understand, no new animations are to be added (that's not what I wrote a brief for at least). It's only adding editing capabilities for properties not already supported in the amp story editor, but supported in amp.

And those are the 5 properties above:

- `animate-in-timing-function`: relevant for all animations.
- `scale-start`: relevant only for "Zoom in" and "Zoom out" animations.
- `scale-end`: relevant only for "Zoom in" and "Zoom out" animations.
- `translate-x`: relevant only for "Pan left" and "Pan right" animations.
- `translate-y`: relevant only for "Pan up" and "Pan down" animations.

I find most of your questions irrelevant in this light.

**Comment by spacedmonkey** [comment]
> I find most of your questions irrelevant in this light.
I disagree. Accessibility is was not even mentioned in your implementation brief.

**Comment by barklund** [comment]
> > I find most of your questions irrelevant in this light.
> 
> I disagree. Accessibility is was not even mentioned in your implementation brief.

**Most**, not all.

It's basically adding a dropdown and some input fields to an existing panel under special circumstances. Defaults are used right now for these values, but not exposed to the editor.

Accessibility is of course relevant for the `BezierEditor` for sure, but as it has input fields as well, I actually think it's covered even though not explicitly mentioned.

**Comment by miina** [comment]
@barklund Thanks for the detailed Implementation Brief.

Questions:
- Do you think it might make sense to just add the pre-defined values for the `animate-in-timing-function` option for now and leave the "custom" for a potential future enhancement?
- On `optionalValues` for Zoom: would there be a way to reset to defaults? Or would the default values be mentioned in the description?
- On Horizontal and Vertical Pan. Looks like the default values are the whole width/height of the specific image. Should there be a "Default" option that would just not add any value to the attribute? Or which kind of input where you thinking of for those two? What did you have in mind for resetting to default?

Note:
- For avoiding the need for [deprecating all the blocks](https://developer.wordpress.org/block-editor/developers/block-api/block-deprecation/) that support animation, we should make sure that new attributes are saved to the post content only if something else than the default is chosen (meaning that for all the previously existing blocks no new attributes would be added and validation errors would not happen).

Thought:
I wonder if it might make sense to add the animation attributes dynamically when rendering (via PHP) instead of saving to the content. This would need migration (deprecating logic) for all the existing blocks, however, it would make it easier to manage for future changes. Perhaps it's not that relevant for this specific issue though and it would be better to address it if it should become necessary.

**Comment by miina** [comment]
Would you mind adding the relevant link(s) from the AMP Stories docs to the Implementation Brief as well for quick access where applicable?

**Comment by barklund** [comment]
@mina, great comments.

> Do you think it might make sense to just add the pre-defined values for the `animate-in-timing-function` option for now and leave the "custom" for a potential future enhancement?

That's most definitely an option. I will ask PO in subsequent comment addressing this specifically.

> On `optionalValues` for Zoom: would there be a way to reset to defaults? Or would the default values be mentioned in the description?

I have not considered resetting to default. A _reset_  button is probably the way to go. Will add to description.

> On Horizontal and Vertical Pan. Looks like the default values are the whole width/height of the specific image. Should there be a "Default" option that would just not add any value to the attribute? Or which kind of input where you thinking of for those two? What did you have in mind for resetting to default?

I was thinking an input field and specifying value in pixels, but maybe a slider from 0 to 200% (of element width/height) makes more sense? Or maybe all to 300%? 400%?

I have not considered resetting to default. A _reset_  button is probably the way to go here as well. Will add to description.

> For avoiding the need for [deprecating all the blocks](https://developer.wordpress.org/block-editor/developers/block-api/block-deprecation/) that support animation, we should make sure that new attributes are saved to the post content only if something else than the default is chosen (meaning that for all the previously existing blocks no new attributes would be added and validation errors would not happen).

Excellent note. I'll make sure to specify.

> I wonder if it might make sense to add the animation attributes dynamically when rendering (via PHP) instead of saving to the content. This would need migration (deprecating logic) for all the existing blocks, however, it would make it easier to manage for future changes. Perhaps it's not that relevant for this specific issue though and it would be better to address it if it should become necessary.

This is beyond me. I have no input here.

**Comment by barklund** [comment]
@MackenzieHartung, what's your thought here?

#### Current Implementation

All animations are linear. There's no option for selecting timing function like easing in or out.

#### Option 1

We add a dropdown with _"Linear"_, _"Ease in"_, _"Ease out"_ and _"Ease in-out"_. Editors can select any one of these and Linear will be default.

**Pro**: Quick implementation, fewer things to consider for editors.
**Con**: Loss of precision for expert users.

#### Option 2

We add a dropdown with _"Linear"_, _"Ease in"_, _"Ease out"_, _"Ease in-out"_ and _"Custom"_. Editors can select any one of these and Linear will be default.

If editor selects Custom, they can input the four timing values used for the cubic bezier function either through for input fields or visually dragging them in an [`BezierEditor` component](https://github.com/gre/bezier-easing-editor).

**Pro**: Full editing capability for expert users.
**Con**: Longer implementation time (+1-2 days), need accessibility consideration, new third party dependency.

**Comment by barklund** [comment]
> Would you mind adding the relevant link(s) from the AMP Stories docs to the Implementation Brief as well for quick access where applicable?

Will do!

**Comment by miina** [comment]
> I was thinking an input field and specifying value in pixels, but maybe a slider from 0 to 200% (of element width/height) makes more sense? Or maybe all to 300%? 400%?

A slider makes sense to me! If I understand the documentation correctly then the value has to be in pixels. If we'd have it in %, would this mean calculation to determine what the % means in pixels? I suppose the default value is the full width/height of the original image?

**Comment by barklund** [comment]
> > I was thinking an input field and specifying value in pixels, but maybe a slider from 0 to 200% (of element width/height) makes more sense? Or maybe all to 300%? 400%?
> 
> A slider makes sense to me! If I understand the documentation correctly then the value has to be in pixels. If we'd have it in %, would this mean calculation to determine what the % means in pixels? I suppose the default value is the full width/height of the original image?

That's an (annoyingly) good question.

I will check exactly what happens now and how we can best support it.

Open questions:

* If we just store the percentage in the block data, when do we do the actual px calculation needed for rendering in PHP?
* And if we store the px result and not percentage, how do we recalculate this (if at all) when the block resizes?

**Comment by miina** [comment]
> If we just store the percentage in the block data, when do we do the actual px calculation needed for rendering in PHP?

For PHP we could do it dynamically right before the rendering. However, note that we also have Animation preview in the editor, meaning that we'd most likely need the final value in the editor, too.

 > And if we store the px result and not percentage, how do we recalculate this (if at all) when the block resizes?

**If** the animation uses the full measures of the original image then we probably would not need to recalculate the values either since in this case, the resizing would be irrelevant.

**Comment by barklund** [comment]
> That's an (annoyingly) good question.
> 
> I will check exactly what happens now and how we can best support it.

So pan is actually a really weird animation. It only works for images as far as I can see and it somewhat ignores the actual image size, and forces it full-screen. It does use the image position, which makes it jump to an unexpected location, as it's resized.

Please compare resulting story here:
https://story-test-wordpress-amp.pantheonsite.io/stories/3493/

With the editor for this story here:
https://story-test-wordpress-amp.pantheonsite.io/wp-admin/post.php?post=3493&action=edit (requires login to Pantheon story test site)

So maybe I actually don't understand these animations.

**Comment by swissspidy** [comment]
Yep, that's how the pan animations work. Animations in AMP are defined here btw: https://github.com/ampproject/amphtml/blob/13b3b6d92ee0565c54ec34732e88f01847aa8a91/extensions/amp-story/1.0/animation-presets.js

For the preview in the editor I basically recreated that logic.

**Comment by miina** [comment]
> So pan is actually a really weird animation. It only works for images as far as I can see and it somewhat ignores the actual image size, and forces it full-screen. It does use the image position, which makes it jump to an unexpected location, as it's resized.

Yep, Pan is definitely an animation that might seem weird at first, it can theoretically be used to create quite interesting effects, especially when using photos prepared for Pan. Eg:

![rabbit](https://user-images.githubusercontent.com/3294597/66509688-d8ee3680-ead3-11e9-8a3d-842cea10bb5a.gif)

Editor:

<img width="335" alt="Screenshot 2019-10-09 at 20 32 42" src="https://user-images.githubusercontent.com/3294597/66509761-f8855f00-ead3-11e9-8395-aea1ffa329af.png">

It's true that the editor and FE differ quite a bit, unless playing the animation.

**Comment by barklund** [comment]
So what would be the relevant range for translate distance? Page width?

> Yep, that's how the pan animations work. Animations in AMP are defined here btw: https://github.com/ampproject/amphtml/blob/13b3b6d92ee0565c54ec34732e88f01847aa8a91/extensions/amp-story/1.0/animation-presets.js
> 
> For the preview in the editor I basically recreated that logic.

Oh, so [0 is the default translate value](https://github.com/ampproject/amphtml/blob/13b3b6d92ee0565c54ec34732e88f01847aa8a91/extensions/amp-story/1.0/animation-presets.js#L274), but as an editor you'd have to set it to `-offsetX`.

Then the range should be values between `-offsetX` and `0`? And as `offsetX` is determined based on the size of the unresized element, it can easily be calculated. And we can then store it as pixels, but let the editor select it in terms of percent, with 100% being default and 0% being no pan.

@swissspidy, does that make sense to you?

**Comment by barklund** [comment]
Here's UI for the implementation suggesting described above:

| Feature | Animation panel |
| - | - |
| Timing function default view | <img width="281" alt="Screenshot 2019-10-09 at 16 07 11" src="https://user-images.githubusercontent.com/637548/66517497-3d4bce80-eab1-11e9-90b6-2f999aa37317.png"> |
| Timing function drop down | <img width="281" alt="Screenshot 2019-10-09 at 16 07 21" src="https://user-images.githubusercontent.com/637548/66517498-3d4bce80-eab1-11e9-86af-f8c1d2466475.png"> |
| Timing function custom | <img width="282" alt="Screenshot 2019-10-09 at 16 14 40" src="https://user-images.githubusercontent.com/637548/66517499-3d4bce80-eab1-11e9-92f1-6c8a64163825.png"> |
| Zoom animation scale settings | <img width="279" alt="Screenshot 2019-10-09 at 16 17 47" src="https://user-images.githubusercontent.com/637548/66517500-3d4bce80-eab1-11e9-8084-8e2cc681f1e0.png"> |
| Pan left/right distance setting | <img width="280" alt="Screenshot 2019-10-09 at 16 22 13" src="https://user-images.githubusercontent.com/637548/66517501-3d4bce80-eab1-11e9-8516-7354e01e932f.png"> |
| Pan up/down distance setting | <img width="281" alt="Screenshot 2019-10-09 at 16 23 00" src="https://user-images.githubusercontent.com/637548/66517502-3de46500-eab1-11e9-8064-64f040350995.png"> |

**Comment by MackenzieHartung** [comment]
I spoke with Vamsee, and the product direction is to go with Option 1. 

> @MackenzieHartung, what's your thought here?
> 
> #### Current Implementation
> All animations are linear. There's no option for selecting timing function like easing in or out.
> 
> #### Option 1
> We add a dropdown with _"Linear"_, _"Ease in"_, _"Ease out"_ and _"Ease in-out"_. Editors can select any one of these and Linear will be default.
> 
> **Pro**: Quick implementation, fewer things to consider for editors.
> **Con**: Loss of precision for expert users.
> 
> #### Option 2
> We add a dropdown with _"Linear"_, _"Ease in"_, _"Ease out"_, _"Ease in-out"_ and _"Custom"_. Editors can select any one of these and Linear will be default.
> 
> If editor selects Custom, they can input the four timing values used for the cubic bezier function either through for input fields or visually dragging them in an [`BezierEditor` component](https://github.com/gre/bezier-easing-editor).
> 
> **Pro**: Full editing capability for expert users.
> **Con**: Longer implementation time (+1-2 days), need accessibility consideration, new third party dependency.

I spoke with Vamsee and the product direction is to go with option 1.

**Comment by swissspidy** [comment]
@barklund Sorry for dropping the ball here.

The relevant range is essentially the element width/height, so what you say makes sense.

> And we can then store it as pixels, but let the editor select it in terms of percent, with 100% being default and 0% being no pan.

What should we do when the user resizes the image so that the stored value becomes greater than the image size? Should we perhaps just store the value in percentages and do the calculation upon serving the story (in PHP)?

**Comment by barklund** [comment]
@swissspidy, please review IB again 😅😅.

**Comment by swissspidy** [comment]
A few minor thoughts:

* With #3421 in mind I don't see an immediate need for storing these additional properties within `timingValues` and `optionalValues` fields. They could  be on the same "level" as the others. 
* > The other options are "Ease in", "Ease out" and "Ease in-out". For the non-linear functions, the following lists of timing values will be used:    
   
   What about "Ease"? It's also mentioned as a default value in the CSS Easing Functions spec. Speaking of which, should we store these as strings rather than the equivalent numbers? Just in case the spec changes. The spec has a similar note:    
    > The keyword values ease, linear, ease-in, ease-out, and ease-in-out are serialized as-is, that is, they are not converted to the equivalent cubic-bezier() function before serializing.
* > Store these if set to non-default values on the optionalValues object as zoomStart and zoomEnd respectively.    
    
    I'd prefer following the naming of  the AMP attributes, e.g. `scaleStart` instead of `zoomStart`.

**Comment by barklund** [comment]
> * With #3421 in mind I don't see an immediate need for storing these additional properties within `timingValues` and `optionalValues` fields. They could  be on the same "level" as the others.

Makes sense and yes for 3421 they should be stored on a similar level. I just didn't want an attribute directly on every element named `translateX`, which would make it difficult to understand what it does. Then it would need a longer name like `ampAnimationTranslateX` or similar.

`timingValues` however is a direct property of the root element. But should probably also be renamed to `ampAnimationTimingValues`.

>   What about "Ease"? It's also mentioned as a default value in the CSS Easing Functions spec. Speaking of which, should we store these as strings rather than the equivalent numbers? Just in case the spec changes. The spec has a similar note:
>   > The keyword values ease, linear, ease-in, ease-out, and ease-in-out are serialized as-is, that is, they are not converted to the equivalent cubic-bezier() function before serializing.

I actually stole them from Lea's cubic-bezier.com. She also has "ease", so it makes sense to add it in there.

I wanted to store them as their value lists for to only have the logic in one place. If we store it as a string, the logic for which values to insert would need to move to the PHP rendering logic. Right?

I don't expect those values to change. These easing calculations have been around forever (at least since [Penner's easing equations for AS1, that I used way back 15+ years ago](http://robertpenner.com/easing/penner_easing_as1.txt))!

>   I'd prefer following the naming of  the AMP attributes, e.g. `scaleStart` instead of `zoomStart`.

As per above it should probably be `ampAnimationScaleStart`.

**Comment by swissspidy** [comment]
> If we store it as a string, the logic for which values to insert would need to move to the PHP rendering logic. Right?

Probably, yes. But looking at [`animate-in-timing-function`](https://amp.dev/documentation/components/amp-story/#animate-in-timing-function-[optional]) again it makes sense to store the values as you proposed 👍
